pkg/venafi/tpp/connector_test.go

```diff
@@ -528,6 +506,73 @@ func TestRequestCertificateWithValidHours(t *testing.T) {
 	DoRequestCertificateWithValidHours(t, tpp)
 }
 
+// When the previous enrolment failed at a stage greater than 0, TPP will return
+// a 500 HTTP error (see example below) when requesting a new certificate. This
+// test ensures that RequestCertificate properly resets the enrolment when this
+// happens.
+//
+// For example, here is the 500 HTTP response body:
+//
+//	{
+//	  "Stage": 700,
+//	  "Status": "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."
+//	}
+//
+// (the Stage number doesn't matter, but it has to be > 0)
+//
+// The reason we are using a mock HTTP server rather than the live TPP server is
+// because consistently triggering the 500 error in a stage different than 0
+// requires putting a powershell script on the TPP VM, which is not something
+// that can be done as part of the "set up" of the current tests we have in
+// vcert.
+func TestRequestCertificateWithReset(t *testing.T) {
+	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t.Logf("request: %s %s", r.Method, r.URL.Path)
+		switch {
+		case r.URL.Path == "/vedsdk/Identity/Self":
+			_, _ = w.Write([]byte("{}"))
+		case r.URL.Path == "/vedsdk/certificates/retrieve":
+			req := struct{ CertificateDN string }{}
+			_ = json.NewDecoder(r.Body).Decode(&req)
+			if req.CertificateDN != `\VED\Policy\Foo\test.com` {
+				t.Fatalf("/vedsdk/certificates/retrieve: expected CertificateDN to be %s but got %s", `\VED\Policy\Foo\test.com`, req.CertificateDN)
+			}
+			w.WriteHeader(500)
```

Comment:
Would it be useful/ possible to also test a 202 case or a 404 case?

Reply:
Yes. I have added three other test cases: one with 200, one with 202, and one with an "unexpected" 500.

Reply:
For context: the "unexpected" 500 is tested in the "should not reset when /certificates/retrieve returns an unrelated 500" function (I think)

Reply:
Yes. I'll make that clearer in the test name.